### PR TITLE
Abstract away the type of tokens in number notations.

### PR DESCRIPTION
**Pull request by ppedrot**
This is a patch extracted from #17881. It makes explicit the invariant we care about when printing number notations. It also makes easier to tinker with the precise algorithm to normalize terms as in #17862.

### CODE REVIEW

**Review by JasonGross — APPROVED** [review]
LGTM (I haven't read through it super carefully, though)

File: interp/notation.ml [hunk 1]
```diff
@@ -770,11 +853,11 @@ let rawnum_of_coquint cl c =
...
-  | App (c,[|c'|]) ->
-     (match Constr.kind c with
-      | Construct ((_,1), _) (* Pos *) -> (SPlus,rawnum_of_coquint cl c')
-      | Construct ((_,2), _) (* Neg *) -> (SMinus,rawnum_of_coquint cl c')
+  match TokenValue.kind c with
+  | TConstruct (c,[c']) ->
+     (match c with
+      | (_, 1) (* Pos *) -> (SPlus,rawnum_of_coquint cl c')
+      | (_, 2) (* Neg *) -> (SMinus,rawnum_of_coquint cl c')
       | _ -> raise NotAValidPrimToken)
```
Comment (proux01) [inline interp/notation.ml:861]:
```suggestion
  match TokenValue.kind c with
  | TConstruct ((_, 1), [c']) (* Pos *) ->SPlus, rawnum_of_coquint cl c'
  | TConstruct ((_, 2), [c']) (* Neg *) ->SMinus, rawnum_of_coquint cl c'
```

File: interp/notation.ml [hunk 2]
```diff
@@ -784,17 +867,14 @@ let rawnum_of_coqnumber cl c =
     let f = try Some (rawnum_of_coquint cl f) with NotAValidPrimToken -> None in
     let e = match e with None -> None | Some e -> Some (rawnum_of_coqint CDec e) in
     NumTok.Signed.of_int_frac_and_exponent n f e in
-  match Constr.kind c with
-  | App (_,[|i; f|]) -> of_ife i f None
-  | App (_,[|i; f; e|]) -> of_ife i f (Some e)
+  match TokenValue.kind c with
+  | TConstruct (_, [i; f]) -> of_ife i f None
+  | TConstruct (_,[i; f; e]) -> of_ife i f (Some e)
```
Comment (proux01) [inline interp/notation.ml:872]:
```suggestion
  | TConstruct (_, [i; f]) -> of_ife i f None
  | TConstruct (_, [i; f; e]) -> of_ife i f (Some e)
```

**Review by proux01 — APPROVED** [review]
LGTM but sounds a bit strange to keep a single `glob_of_constr` use on line 1059

**Comment by proux01** [comment]
Ok, it currently feels duplicated to me but I'm certainly missing something.

**Pull request merged**